### PR TITLE
Added ability to copy-paste background elements

### LINKED ISSUE #1440: Copy/paste the background layer should not create new layers
## Bug Description

fun bug: I can copy the background and then paste it a couple of times to have unlimited background layers

## Expected Behaviour

Copy/pasting the background layer / page should replace the background layer in the page it is pasted. 
1. If pasted on the same page, in essence it does nothing. 
2. If pasted in another page, replace that page's background. 

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### Copying
* When a shape background is copied, make sure to get the page background color from the page and move along with shape to clipboard (post-merge of #2158 moving background color from element to page)
* When a media background is copied, no extra attributes need to be copied (post-merge of #1925 moving background overlay from page to element).

### Pasting
* When pasting a background shape element, only look at the background element color at make sure to set it as the page background color. However, also remove (delete) any current background media element (to reveal the "raw" background shape element).
* When pasting a background media element, make sure to merge it into the current background using the "combineElements" operation (will make sure it happens just like regular merge), except also make sure that background overlay is correctly handled.

### Copy shape to another shape
1. Create a story with two pages - with two different background colors (can be gradients)
1. Select the background element from page 1 and press <kbd>mod</kbd>+<kbd>C</kbd> to copy
1. Switch to page 2 and press <kbd>mod</kbd>+<kbd>V</kbd> to paste
1. Observe that background element on page 2 is now identical to page 1 and page 2 still only has a single layer (background)
1. Press "undo" (shortcut or button) and observe that entire operation can be cancelled in a single undo step

### Copy shape to media
1. Create a story with two pages - page 1 has a color background and page 2 has an image background
1. Select the background element from page 1 and press <kbd>mod</kbd>+<kbd>C</kbd> to copy
1. Switch to page 2 and press <kbd>mod</kbd>+<kbd>V</kbd> to paste
1. Observe that background element on page 2 is now identical to page 1 (background media element has been deleted) and page 2 still only has a single layer (background)
1. Press "undo" (shortcut or button) and observe that entire operation can be cancelled in a single undo step

### Copy media to shape
1. Create a story with two pages
1. Set a red background color on page 1, add an image, set it as background and add any overlay to the image
1. Set a blue background color on page 2
1. Select the background element from page 1 and press <kbd>mod</kbd>+<kbd>C</kbd> to copy
1. Switch to page 2 and press <kbd>mod</kbd>+<kbd>V</kbd> to paste
1. Observe that background element on page 2 is now identical to page 1 (background media element has been inserted with overlay) and page 2 still only has a single layer (background)
1. Also observe, that page 2 still has a blue page background color "behind" the media element, whereas page 1 still has the red page background color
1. Press "undo" (shortcut or button) and observe that entire operation can be cancelled in a single undo step

### Copy media to media
1. Create a story with two pages
1. Set a red background color on page 1, add an image, set it as background and add any overlay to the image
1. Set a blue background color on page 2, add another image, set it as background and add any overlay (different from page 1) to the image
1. Select the background element from page 1 and press <kbd>mod</kbd>+<kbd>C</kbd> to copy
1. Switch to page 2 and press <kbd>mod</kbd>+<kbd>V</kbd> to paste
1. Observe that background element on page 2 is now identical to page 1 (background media element has been swapped and has a new overlay) and page 2 still only has a single layer (background)
1. Also observe, that page 2 still has a blue page background color "behind" the media element, whereas page 1 still has the red page background color
1. Press "undo" (shortcut or button) and observe that entire operation can be cancelled in a single undo step

### Duplicate background element
1. Create a story with any background (shape or media)
1. Duplicate background element by pressing <kbd>mod</kbd>+<kbd>D</kbd>
1. Observe that nothing happens

### LINKED ISSUE #2709: Copy Paste Not Working on Windows
## Bug Description

Per Claudio on #1440:

> Testing in QA Env. - PR#2387
> 
> Tested on a PC so using ctrl-c/ctril-v but it isn't working (unable to test on a mac at the moment). Unable to cut-and-paste

## Additional Context

- Plugin Version:
- Operating System:
- Browser:

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by barklund**
## Summary

This modifies both how elements are copied (background shapes are added page background color), and how elements are pasted (pasted background elements are filtered out and handled separately).

## To-do

- [x] Modify copying/cutting background shapes (add page color to element)
- [x] Modify how background elements are handled on paste (separate handling)
- [x] Wait for both dependent PR's to be merged (#1925, ~#2158~)
- [x] Update and amend unit tests for copy-paste
- [x] Add karma test
- [x] Also handle duplication
- [x] Add karma test for duplication

## User-facing changes

Copy-pasting elements now does not allow a user to make multiple copies of a background element (shape or media) - if you paste a background element, it replaces what was already background on the given page - thus completely removing what was there.

If you copy and paste a bg to the same page, nothing happens (actually the bg element is updated, but to something identical to itself).

## Testing Instructions

1. Select the background element on any page (shape or image)
1. Copy the element via keyboard shortcut
1. Paste via keyboard shortcut
1. Observe that nothing happens.

For full QA testing instructions, please see #1440.

## Fixes

Fixes #1440
Fixes #2709

### CODE REVIEW

**Review by dvoytenko — APPROVED** [review]

**Review by miina — CHANGES_REQUESTED** [review]
Looks like handling for cloning the background is missing (see `cloneHandler` in `useCanvasKeys`).

**Review by merapi — APPROVED** [review]
Works as expected for shapes and images, waiting for @samitron7 to confirm how video BG should behave (probably should autoplay like when dropped), either way, looks like we will need a small fix here (right now the pause button is displayed but the video is not playing), I can do it, should I fix it here or in another PR?

**Comment by barklund** [comment]
> Looks like handling for cloning the background is missing

@miina, good catch, implementing now (with tests).

> I can do it, should I fix it here or in another PR?

@merapi, let's file that as a follow-up.

File: assets/src/edit-story/components/canvas/useAddNewElements.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    [
+      addElements,
+      currentPage,
+      updateCurrentPageProperties,
+      combineElements,
+      deleteElementById,
+    ]
+  );
+
+  return addNewElements;
```
Comment (miina) [inline assets/src/edit-story/components/canvas/useAddNewElements.js]: I wonder if there's a way to better distinguish `addNewElements` from `addElements`. `addNewElements` suggests as if `addElements` would perhaps add existing elements, actually it's always kind of adding new elements. Or well, if the elements are copied-pasted then one might argue that `addNewElements` is actually the one adding the old elements 😛 

Anyway -- wondering if there's a better name for this. It looks like `addNewElements` is always used for the pasted elements only, is that correct?

File: assets/src/edit-story/components/canvas/karma/cloneSelection.karma.js [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+  afterEach(() => {
+    fixture.restore();
+  });
+
+  it('should correctly do nothing if no selection', async () => {
+    // Clear selection by pressing above background element
+    await clickAboveElement(bg.id);
+
+    //await karmaPause();
```
Comment (miina) [inline assets/src/edit-story/components/canvas/karma/cloneSelection.karma.js]: Can be removed?

File: assets/src/edit-story/components/canvas/karma/cloneSelection.karma.js [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+      const actual = element?.[property] ?? '';
+      const pass =
+        typeof expected === 'string'
+          ? util.equals(actual, expected, customEqualityTesters)
+          : expected.test(actual);
+      return {
+        pass,
+        message: pass
+          ? `Expected element to not have src "${expected}"`
+          : `Expected element to have src "${expected}" but found "${actual}"`,
```
Comment (miina) [inline assets/src/edit-story/components/canvas/karma/cloneSelection.karma.js:252]: ^ Is using "src" intended here, for the general `toHaveProperty`? Is guess in this context it's OK if it's always the src you're comparing.

File: assets/src/edit-story/components/canvas/karma/backgroundCopyPaste.karma.js [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+      const actual = element?.[property] ?? '';
+      const pass =
+        typeof expected === 'string'
+          ? util.equals(actual, expected, customEqualityTesters)
+          : expected.test(actual);
+      return {
+        pass,
+        message: pass
+          ? `Expected element to not have src "${expected}"`
+          : `Expected element to have src "${expected}" but found "${actual}"`,
```
Comment (miina) [inline assets/src/edit-story/components/canvas/karma/backgroundCopyPaste.karma.js]: Same question here, should the `src` actually be `${property}`?

**Review by miina — APPROVED** [review]
Added a few nit-comments but generally LGTM!

File: assets/src/edit-story/components/canvas/useAddNewElements.js (hunk 1 above)
Comment (barklund) [reply assets/src/edit-story/components/canvas/useAddNewElements.js]: Yeah, was coming up short with a name here. `addPastedElements`?

Comment (miina) [reply assets/src/edit-story/components/canvas/useAddNewElements.js]: SGTM!

File: assets/src/edit-story/components/canvas/karma/cloneSelection.karma.js (hunk 2 above)
Comment (barklund) [reply assets/src/edit-story/components/canvas/karma/cloneSelection.karma.js]: Fixed in 091fad534.

File: assets/src/edit-story/components/canvas/karma/cloneSelection.karma.js (hunk 3 above)
Comment (barklund) [reply assets/src/edit-story/components/canvas/karma/cloneSelection.karma.js:252]: Fixed in 091fad534.

File: assets/src/edit-story/components/canvas/karma/backgroundCopyPaste.karma.js (hunk 4 above)
Comment (barklund) [reply assets/src/edit-story/components/canvas/karma/backgroundCopyPaste.karma.js]: Fixed in 091fad534.

File: assets/src/edit-story/components/canvas/useAddNewElements.js (hunk 1 above)
Comment (barklund) [reply assets/src/edit-story/components/canvas/useAddNewElements.js]: Renamed in 091fad534.

File: assets/src/edit-story/components/canvas/karma/cloneSelection.karma.js [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    // Drag it to (50,50)
+    const rect1 = (
+      await getCanvasElementWrapperById(img1.id)
+    ).getBoundingClientRect();
+    await fixture.events.mouse.seq(({ move, down, up }) => [
+      move(rect1.left + 10, rect1.top + 10),
+      down(),
+      move(bgRect.left + 50, bgRect.top + 50, { steps: 2 }),
+      up(),
+    ]);
```
Comment (barklund) [inline assets/src/edit-story/components/canvas/karma/cloneSelection.karma.js:49]: The need for dragging here is to be able to correctly select the items later by clicking them in the top left corner - they need to not overlap there (and as they're currently placed randomly on insertion, it failed sporadically when they happened to overlap).

**Comment by merapi** [comment]
@googlebot I consent.

**Pull request merged**